File: 2015/day18.js
let grid = require('./input/day18')
  .split("\n")
  .map(line => (
    line.split('')
        .map(char => char === '#' ? true : false)
  ))
const length = grid[0].length - 1
const height = grid.length - 1
grid[0][0] = true
grid[0][length] = true
grid[height][0] = true
grid[height][length] = true
let test = `##.#.#
...##.
#....#
..#...
#.#..#
####.#`
  .split("\n")
  .map(line => (
    line.split('')
        .map(char => char === '#' ? true : false)
  ))
const neighborCount = (total, curr) => curr ? ++total : total

const sliceBoundaries = (length, j) => [Math.max(j - 1, 0), Math.min(length, j + 2)]

const sideNeighbors = (arr, j) => (
  (arr[j - 1] ? 1 : 0) +
  (arr[j + 1] ? 1 : 0) 
)

const lowNeighbors = (grid, arr, j) => (
  sideNeighbors(arr, j) +
  grid[1].slice(...sliceBoundaries(arr.length, j))
    .reduce(neighborCount, 0)
)

const topNeighbors = (grid, arr, j) => (
  sideNeighbors(arr, j) +
  grid[grid.length - 2].slice(...sliceBoundaries(arr.length, j))
    .reduce(neighborCount, 0)
)

const midNeighbors = (grid, arr, i, j) => (
  sideNeighbors(arr, j) +
  grid[i - 1].slice(...sliceBoundaries(arr.length, j))
  .concat(
    grid[i + 1].slice(...sliceBoundaries(arr.length, j))
  ).reduce(neighborCount, 0)
)

const neighbors = (grid, arr, i, j) => (
  i === 0 ? lowNeighbors(grid, arr, j) :
  i === grid.length - 1 ? topNeighbors(grid, arr, j) :
  midNeighbors(grid, arr, i, j)
)

const turnOn = (grid, arr, char, i, j) => (
  !char && neighbors(grid, arr, i, j) === 3
)

const stayOn = (grid, arr, char, i, j) => (
  char && [2, 3].includes(neighbors(grid, arr, i, j))
)

const corner = (i, j, height, length) => (
  (i === 0 && (j === 0 || j === length)) || 
  (i === height && (j === 0 || j === length))
)

const next = grid => (
  grid.map((arr, i) => (
    arr.map((char, j) => (
      turnOn(grid, arr, char, i, j) || 
      stayOn(grid, arr, char, i, j)) || 
      corner(i, j, grid.length - 1, arr.length - 1)
    )
  ))
)

for(let i = 0; i <= 100; i++) {
  console.log(i)
  console.log(grid.reduce((tot, row) => tot + row.reduce(neighborCount, 0), 0))
  //console.log(test.map(row => row.map(char => char ? '#' : '.', '').join('')).join("\n"))
  grid = next(grid)
}

File: 2015/day18.rb
input = File.read('./input/day18.txt')

class Grid
  def initialize(input)
    @cells = input.split("\n").map{|line| line.split('').map{ |c| c == '#' }}
  end

  def step
    @cells = @cells.each_with_index.map{|line, i| line.each_with_index.map{|c, j| next_char(c, i, j) }}
  end

  def next_char(c, i, j)
    top = i == 0 ? [] : @cells[i - 1]
    bottom = i == @cells.length - 1 ? [] : @cells[i + 1]
    left = j == 0 ? nil : j - 1
    right = j == @cells[0].length - 1 ? nil : j + 1
    neighbors = [(top[left] if left), top[j], (top[right] if right),
     (@cells[i][left] if left), (@cells[i][right] if right),
     (bottom[left] if left), bottom[j], (bottom[right] if right)].reject{|c| !c }.length
    neighbors == 3 || (c && neighbors == 2)
  end

  def to_s
    @cells.map{|line| line.map{|c| c ? '#' : '.'}.join}.join("\n")
  end

  def count
    @cells.reduce(0){|total, line| line.reject{|c| !c}.length + total }
  end
end

test_data = '.#.#.#
...##.
#....#
..#...
#.#..#
####..'
test = Grid.new(test_data)
5.times{ test.step; puts test }
actual = Grid.new(input)
100.times{ actual.step }
puts actual.count

class CornerGrid < Grid
  def initialize(input)
    super(input)
    turn_on_corners
  end

  def step
    super
    turn_on_corners
  end
 
  def turn_on_corners
    last = @cells[0].length - 1
    @cells[0][0] = true
    @cells[0][last] = true
    @cells[@cells.length - 1][0] = true
    @cells[@cells.length - 1][last] = true
  end
end

corner = CornerGrid.new(input)
100.times { corner.step }
puts corner.count
